pallets/loans/src/entities/pricing/external.rs

```diff
@@ -94,36 +125,44 @@ impl<T: Config> ExternalActivePricing<T> {
 	) -> Result<T::Balance, DispatchError> {
 		match self.info.max_borrow_amount {
 			MaxBorrowAmount::Quantity(quantity) => {
-				let price = self.calculate_price()?;
+				let price = self.current_price()?;
 				let available = quantity.ensure_sub(self.outstanding_quantity)?;
 				Ok(price.ensure_mul_int(available)?)
 			}
 			MaxBorrowAmount::NoLimit => Ok(desired_amount),
 		}
 	}
 
-	pub fn last_updated(&self) -> Result<Moment, DispatchError> {
-		Ok(T::PriceRegistry::get(&self.info.price_id)?.1)
-	}
+	pub fn adjust(
+		&mut self,
+		principal_adj: Adjustment<T::Balance>,
+		interest: T::Balance,
+	) -> DispatchResult {
+		let quantity_adj = principal_adj.try_map(|principal| -> Result<_, DispatchError> {
+			let price = self.current_price()?;
 
-	pub fn compute_present_value(&self, price: T::Rate) -> Result<T::Balance, DispatchError> {
-		Ok(price.ensure_mul_int(self.outstanding_quantity)?)
-	}
+			let quantity = T::Rate::saturating_from_integer(principal)
```

Comment:
I am trying to understand what we are doing here. 

---
* any letter or number with an elvated annotation of $DIV$, eg. $x^{DIV}$, should be a number in the decimals of `T::Rate`
* $DIV$ is just $10^{decimals}$
---
What we have here given we want to adjust the principal by `100`. Assuming pool currency has NO decimals.
* $p$ is the price
* quantitiy: $q = \frac{100 * DIV}{p^{DIV}} * DIV$

And we are checking at the end in the ensure clause:
$p^{DIV} * q = 100$ &rarr; $p^{DIV} * \frac{100 * DIV}{p^{DIV}} * DIV = 100$

where `ensure_mul_int` is NOT adapting the decimals of `100`.

Reply:
I am either not writing this down correctly (probably) or something is off here ^^

Reply:
Let me check this deeply 👍🏻

Reply:
Well, first, some questions/considerations about the types:
- Currently, `Price` is a `Rate`, but should it be? what should be a `Balance` and what should be a `Rate`? still not clear to me.
- Quantity is a `Balance`, but it's just a multiplier to that price, maybe it deserves another type.

Given said that and supposing 2-decimal precision, if my price is `2000.01` and my quantity is 5, my `Balance` as a result of multiplying this should be `1000005`, right?

---

> Will the `principa_adj` already have the right decimals?

Whatever `principal` is, comes from multiplying the current price by quantity on the client side. So it should have the right decimals if the client side did it well (we should probably create an RPC for this). i.e. if the price is `2000.01` and quantity 5, it should be `10000.05` that interpreted as balance is 1000005

> I think this should rather be `T::Rate::from_inner()`

Still need to think more about this, but I think you're right. 
Testing is passing, but I think I'm making a bad assumption calculating the principle as follows: 
```rust		
let amount = PRICE_VALUE.saturating_mul_int(QUANTITY);
```
which makes me lose the price decimals.

> Given we do not want to saturate here, or do we want to?

I think we should check it, yes. We do not want to saturate here 👍🏻 

> * quantitiy: $q = \frac{100 * DIV}{p^{DIV}} * DIV$

Why not just $q = \frac{100 * DIV}{p^{DIV}}$ without the last `DIV`? from where does the last `DIV` come?

> where `ensure_mul_int` is NOT adapting the decimals of 100

Not sure what you mean here. `principal` is a `Balance`, so any "decimal" is part of the integer value itself that it is multiplying by the rate.

Reply:
I do not know if I have enough information from the abstract type to do the correct maths.

What I need is an integer `Balance` (which I do not know the precision convention) as a result of a multiplication of a `Rate` by a multiplier. How do I know how many decimals from the price I should keep for the `Balance` result?

Reply:
When would the u128 actually overflow?

I'm going to check it. 

Anyways I think from the `pallet-loans` perspective, I have no way to be sure if the user is using a `Rate` with more precision than a `Balance`. I think the big issue here is that we are mixing `Balance` with `Rate` to represent the "same" thing. Maybe we need to choose how to represent any kind of currency value in `Balance` or `Rate` and not mix them... 🤔

Reply:
We've definitely ended up using `Rate` where we shouldn't. This has generally been done where we want to keep working in the 27-decimal fixed-point space rather than keeping integers around (I don't believe we have a different balance type).

Ideally, we'd be using `Balance` anywhere we represent a quantity, and only use Rate for, well, rates. Interest accrual, penalties, etc should all be done in the `Rate` space, and then multiplied by the integer `Balance`

In this case, since `quantity` is also a Balance (though maybe we should make that a different integer type, since it doesn't represent a currency amount?), I think we don't need any specific fancy fixed-point math, or even checked math:

```
quantity = principal / price;
ensure!(principal % price == 0)
```

Reply:
Thanks @branan, for coming to the conversation. I think you've pointed out the origin of the issue correctly. 

Actually, `Price` was a `Balance` initially, but we changed after seeing other chains using rates for that in #1387 (but if I remember well @offerijns, we had both no strong opinion on doing it, just seemed more "appropriated"). Seeing now how a `Rate` price behaves now with the rest of the amount values makes much more sense to change it again to `Balance` and avoid all these issues.

Reply:
> since quantity is also a Balance (though maybe we should make that a different integer type, since it doesn't represent a currency amount?)

Yes, it doesn't represent a currency amount, only a multiplier. However, you could see this as if this quantity is, in fact, the number of "tokens" of a T-bill currency.

Reply:
For quantity, I do think we want some decimal precision ideally (so you can have a quantity of 0.1). So would probably keep that as a Balance.

For changing price to a Balance, what would the decimals of that then be though? Right now, the decimals of all Balances are always based on the pool currency, while Rate is a fixed 27 decimals. The prices can be used by multiple pools so cannot be linked to 1 pool currency. I still don't really see why changing price to a Balance would solve the above issue, but I might not fully follow it.

Reply:
If we need to compute the result as Balance, we do not know how many decimals from the Rate I should choose.

Nevertheless, having price as a Balance, no matter the precision the user chooses, it will be the same for both the price and the amount. The oracle should be responsable from setting a correct price with the required Balance format.

Reply:
Now I think, that Quantity should be a Rate, because it acts as a multiplier of another type.

Reply:
> Right now, the decimals of all Balances are always based on the pool currency

Could you expand this @offerijns? Are we using different Balances with different decimal precision?

Reply:
Yes all `Balance` properties in a pool (e.g. all invested amounts, the reserve, ...) as well as all the loan properties (e.g. the collateral value, borrowed amount, ...) are based on the decimal precision of the pool currency.

Reply:
Will be here: https://github.com/centrifuge/centrifuge-chain/issues/1447